### PR TITLE
Use exchange to send CASE session establishment messages

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem

CASE messages are sent directly using SecureSessionMgr, which is UDP based. It should instead use Exchange and CRMP.

 #### Summary of Changes

Move CASESession code to protocols/secure_channel, and use Exchange and CRMP to send and receive the messages.

### CODE REVIEW

File: src/protocols/secure_channel/CASESession.cpp [hunk 1]
```diff
@@ -283,7 +296,7 @@ CHIP_ERROR CASESession::DeriveSecureSession(const uint8_t * info, size_t info_le
 
 CHIP_ERROR CASESession::SendSigmaR1()
 {
-    uint16_t data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) +
+    uint16_t data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) + sizeof(uint16_t) +
```
Comment (andy31415) [inline src/protocols/secure_channel/CASESession.cpp:299]: could we extract the sigma1 encoding into some separate file with unit tests? These sizeof calculations don't seem obvious to double-check

**Review by andy31415 — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.h [hunk 2]
```diff
@@ -101,13 +105,14 @@ class DLL_EXPORT CASESession
...
      * @param myKeyId                       Key ID to be assigned to the secure session on the peer node
+     * @param exchangeCtxt                  The exchange context to send and receive messages with the peer
      * @param delegate                      Callback object
      *
      * @return CHIP_ERROR      The result of initialization
      */
     CHIP_ERROR EstablishSession(const Transport::PeerAddress peerAddress, OperationalCredentialSet * operationalCredentialSet,
                                 Optional<NodeId> myNodeId, NodeId peerNodeId, uint16_t myKeyId,
-                                SessionEstablishmentDelegate * delegate);
+                                Messaging::ExchangeContext * exchangeCtxt, SessionEstablishmentDelegate * delegate);
```
Comment (msandstedt) [inline src/protocols/secure_channel/CASESession.h:115]: Why do we pass `myNodeId` when we are already passing `operationalCredentialSet`?  This seems inherently redundant.  What happens if `myNodeId` is different than the authenticated RDN in `operationalCredentialSet`?

File: src/protocols/secure_channel/CASESession.cpp (hunk 1 above)
Comment (pan-apple) [reply src/protocols/secure_channel/CASESession.cpp:299]: The message construction and parsing will be changed to TLV formatting. So this code will get reworked. I agree that we should extract out the logic to help unit test it.

**Review by msandstedt — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.cpp [hunk 3]
```diff
@@ -251,6 +248,16 @@ CHIP_ERROR CASESession::EstablishSession(const Transport::PeerAddress peerAddres
...
 }
 
+void CASESession::OnResponseTimeout(ExchangeContext * ec)
+{
+    VerifyOrReturn(ec != nullptr, ChipLogError(Inet, "CASESession::OnResponseTimeout was called by null exchange"));
+    VerifyOrReturn(mExchangeCtxt == nullptr || mExchangeCtxt == ec,
+                   ChipLogError(Inet, "CASESession::OnResponseTimeout exchange doesn't match"));
+    ChipLogError(Inet, "CASESession timed out while waiting for a response from the peer. Expected message type was %d",
+                 mNextExpectedMsg);
+    mDelegate->OnSessionEstablishmentError(CHIP_ERROR_TIMEOUT);
```
[the remark this answers is not in the document]
Comment (pan-apple) [reply src/protocols/secure_channel/CASESession.cpp:258]: Yes, that's a good catch. I'll add Clear() here.

File: src/protocols/secure_channel/CASESession.cpp [hunk 4]
```diff
@@ -251,6 +248,16 @@ CHIP_ERROR CASESession::EstablishSession(const Transport::PeerAddress peerAddres
     return err;
 }
 
+void CASESession::OnResponseTimeout(ExchangeContext * ec)
+{
+    VerifyOrReturn(ec != nullptr, ChipLogError(Inet, "CASESession::OnResponseTimeout was called by null exchange"));
+    VerifyOrReturn(mExchangeCtxt == nullptr || mExchangeCtxt == ec,
```
[the remark this answers is not in the document]
Comment (pan-apple) [reply src/protocols/secure_channel/CASESession.cpp:254]: I cannot think of a scenario where the response timeout will be called until we sent a message first, using an exchange. I'll remove this check.

File: src/protocols/secure_channel/CASESession.cpp [hunk 5]
```diff
@@ -1068,51 +1087,55 @@ void CASESession::HandleErrorMsg(const PacketHeader & header, const System::Pack
...
+void CASESession::OnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, const PayloadHeader & payloadHeader,
+                                    System::PacketBufferHandle msg)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
-    PayloadHeader payloadHeader;
-
-    Protocols::SecureChannel::MsgType msgType;
 
+    VerifyOrExit(ec != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mExchangeCtxt == nullptr || mExchangeCtxt == ec, err = CHIP_ERROR_INVALID_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (pan-apple) [reply src/protocols/secure_channel/CASESession.cpp:1096]: `mExchangeCtxt` can be null here if the `OnMessageReceived` was called by UMH. I'll document it.

The `mExchangeCtxt` is closed during Clear(). Clear() is called during most of the error handlings.

I think one condition where it's not cleared is when `(mExchangeCtxt != nullptr && mExchangeCtxt != ec)`. Ideally it shouldn't happen, but I should still add the code to close the incoming `ec`.

[the remark this answers is not in the document]
Comment (pan-apple) [reply src/protocols/secure_channel/CASESession.cpp:1096]: Actually, come to think of it, the initial error checks in `CASESession::OnMessageReceived` will not close the exchange. I'll refactor it out to cleanly close it.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**